### PR TITLE
ENG-8191: Add configurable export push interval.

### PR DESCRIPTION
**Pull request by akhanzode**

### CODE REVIEW

File: examples/voter/deployment.xml [hunk 1]
```diff
@@ -4,4 +4,7 @@
     <httpd enabled="true">
         <jsonapi enabled="true" />
     </httpd>
+    <systemsettings>
+        <exportPushInterval interval="1000"/>
```
Comment (akhanzode) [inline examples/voter/deployment.xml]: oops I will undo this.

File: src/catgen/spec.txt [hunk 2]
```diff
@@ -34,6 +34,7 @@ begin Systemsettings javaonly "Container for deployment systemsettings element"
   int elasticduration         "Maximum duration time for rebalancing"
   int elasticthroughput       "Target throughput in megabytes for elasticity"
   int querytimeout            "The maximum latency for a query batch before timing out"
+  int exportpushinterval      "Minimum time to wait before export push is performed"
```
Comment (nshi) [inline src/catgen/spec.txt]: **Max** instead of min?

File: src/ee/execution/VoltDBEngine.cpp [hunk 3]
```diff
@@ -1430,24 +1432,26 @@ typedef std::pair<std::string, Table*> TablePair;
 /** Perform once per second, non-transactional work. */
 void VoltDBEngine::tick(int64_t timeInMillis, int64_t lastCommittedSpHandle) {
     m_executorContext->setupForTick(lastCommittedSpHandle);
+    int32_t interval = m_executorContext->getEngine()->exportPushInterval();
```
Comment (nshi) [inline src/ee/execution/VoltDBEngine.cpp]: Why not just call `exportPushInterval()` directly in the loop below?

File: src/ee/execution/VoltDBEngine.cpp [hunk 4]
```diff
@@ -1430,24 +1432,26 @@ typedef std::pair<std::string, Table*> TablePair;
...
     if (m_drReplicatedStream) {
-        m_drReplicatedStream->periodicFlush(timeInMillis, lastCommittedSpHandle);
+        m_drReplicatedStream->periodicFlush(timeInMillis, lastCommittedSpHandle, interval);
     }
 }
 
 /** For now, bring the Export system to a steady state with no buffers with content */
 void VoltDBEngine::quiesce(int64_t lastCommittedSpHandle) {
     m_executorContext->setupForQuiesce(lastCommittedSpHandle);
+    int32_t interval = m_executorContext->getEngine()->exportPushInterval();
```
Comment (nshi) [inline src/ee/execution/VoltDBEngine.cpp]: Same here.

File: src/frontend/org/voltdb/compiler/DeploymentFileSchema.xsd [hunk 5]
```diff
@@ -131,6 +131,13 @@
     </xs:restriction>
   </xs:simpleType>
 
+  <!-- minimum push interval is 1/2 second -->
+  <xs:simpleType name="exportPushIntervalType">
+    <xs:restriction base="xs:int">
+      <xs:minInclusive value="500"/>
```
Comment (nshi) [inline src/frontend/org/voltdb/compiler/DeploymentFileSchema.xsd]: Should probably match this with the Site tick interval, which is 5ms.

File: src/frontend/org/voltdb/iv2/Site.java [hunk 6]
```diff
@@ -494,8 +497,8 @@ ExecutionEngine initializeEE()
                         m_partitionId,
                         CoreUtils.getHostIdFromHSId(m_siteId),
                         hostname,
-                        m_context.cluster.getDeployment().get("deployment").
-                        getSystemsettings().get("systemsettings").getTemptablemaxsize(),
+                        ss.get("systemsettings").getTemptablemaxsize(),
```
Comment (nshi) [inline src/frontend/org/voltdb/iv2/Site.java]: You can get the system settings object above and reuse it here.

File: src/ee/execution/VoltDBEngine.cpp [hunk 7]
```diff
@@ -1431,11 +1433,11 @@ typedef std::pair<std::string, Table*> TablePair;
 void VoltDBEngine::tick(int64_t timeInMillis, int64_t lastCommittedSpHandle) {
     m_executorContext->setupForTick(lastCommittedSpHandle);
     BOOST_FOREACH (TablePair table, m_exportingTables) {
-        table.second->flushOldTuples(timeInMillis);
+        table.second->flushOldTuples(timeInMillis, m_executorContext->getEngine()->exportPushInterval());
```
Comment (vtkstef) [inline src/ee/execution/VoltDBEngine.cpp]: Shouldn't it be a context read only variable

File: src/frontend/org/voltdb/RealVoltDB.java [hunk 8]
```diff
@@ -1467,6 +1467,9 @@ int readDeploymentAndCreateStarterCatalogContext() {
                 if (sysType.getQuery() != null && sysType.getQuery().getTimeout() > 0) {
                     hostLog.info("Query timeout set to " + sysType.getQuery().getTimeout() + " milliseconds");
                 }
+                if (sysType.getExportPushInterval() != null && sysType.getExportPushInterval().getInterval() > 0) {
```
Comment (vtkstef) [inline src/frontend/org/voltdb/RealVoltDB.java]: probably 1 millis is too low we should have a lower bound of say 100 and an upper bound of 4000

File: src/frontend/org/voltdb/compiler/DeploymentFileSchema.xsd [hunk 9]
```diff
@@ -155,6 +162,11 @@
                 <xs:attribute name="timeout" type="latencyType" default="0"/>
             </xs:complexType>
         </xs:element>
+        <xs:element name="exportPushInterval" minOccurs="0" maxOccurs="1">
```
Comment (vtkstef) [inline src/frontend/org/voltdb/compiler/DeploymentFileSchema.xsd]: wouldn't `export-push-interval` be a better name? Ask AJ please

File: tests/ee/execution/add_drop_table.cpp [hunk 10]
```diff
@@ -59,7 +59,7 @@ class AddDropTableTest : public Test {
                              m_hostId,
                              m_hostName,
                              false,
-                             DEFAULT_TEMP_TABLE_MEMORY);
+                             DEFAULT_TEMP_TABLE_MEMORY, 4000);
```
Comment (vtkstef) [inline tests/ee/execution/add_drop_table.cpp]: add a constant DEFAULT_EXPORT_PUSH_INTERVAL

File: tests/ee/execution/engine_test.cpp [hunk 11]
```diff
@@ -146,7 +146,7 @@ class ExecutionEngineTest : public Test {
              */
             engine = new VoltDBEngine();
             int partitionCount = 3;
-            ASSERT_TRUE(engine->initialize(this->cluster_id, this->site_id, 0, 0, "", false, DEFAULT_TEMP_TABLE_MEMORY));
+            ASSERT_TRUE(engine->initialize(this->cluster_id, this->site_id, 0, 0, "", false, DEFAULT_TEMP_TABLE_MEMORY, 4000));
```
Comment (vtkstef) [inline tests/ee/execution/engine_test.cpp]: same here

**Comment by vtkstef** [comment]
Use ExecutorContext instead of weaving this parameter everywhere

**Comment by akhanzode** [comment]
@vtkstef I am not going to merge this branch to master until we talk to cust and run some tests with various settings. I will address the comments. Also need to get Andrews review before we merge this to master. Thanks for looking at this PR

File: src/frontend/org/voltdb/compiler/DeploymentFileSchema.xsd [hunk 12]
```diff
@@ -131,6 +131,13 @@
     </xs:restriction>
   </xs:simpleType>
 
+  <!-- minimum push interval is 1/2 second -->
```
Comment (nshi) [inline src/frontend/org/voltdb/compiler/DeploymentFileSchema.xsd]: Did you mean 5ms?

**Pull request closed without merging**